### PR TITLE
feat(media): add multimodal support

### PR DESCRIPTION
**Pull request by hassiebp**
> [!IMPORTANT]
> Adds multimodal support to Langfuse with new media handling class, API updates, and enhanced OpenAI integration.
> 
>   - **Media Handling**:
>     - Introduces `LangfuseMedia` class for handling media content, supporting base64 data URIs and raw content bytes.
>     - Adds methods for processing media in events, including uploading and patching media records.
>   - **API Changes**:
>     - Updates `server.ts` to include new endpoints for media operations: `media_get`, `media_patch`, and `media_getUploadUrl`.
>     - Adds new schemas in `server.ts` for media-related requests and responses.
>   - **Testing**:
>     - Adds `langfuse.media.spec.ts` to test `LangfuseMedia` functionalities, including constructor, parsing, and JSON serialization.
>   - **OpenAI Integration**:
>     - Updates `observeOpenAI.ts` to include `shutdownAsync` method for Langfuse client.
>     - Enhances OpenAI SDK tracing to support media handling.
>   - **Miscellaneous**:
>     - Updates `types.ts` to include new types for media operations.
>     - Minor import and type adjustments in `LangfuseExporter.ts` and `types.ts`.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=langfuse%2Flangfuse-js&utm_source=github&utm_medium=referral)<sup> for 356b4facb755f5ddddf31778954a63516fa1340b. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -70,11 +70,11 @@ jobs:
           cd ./langfuse-server
 
           echo "::group::Run langfuse server"
-          TELEMETRY_ENABLED=false docker compose up -d db
+          TELEMETRY_ENABLED=false docker compose -f docker-compose.v3preview.yml up -d postgres
           echo "::endgroup::"
 
           echo "::group::Logs from langfuse server"
-          TELEMETRY_ENABLED=false docker compose logs
+          TELEMETRY_ENABLED=false docker compose -f docker-compose.v3preview.yml logs
```
Comment (maxdeichmann) [inline .github/workflows/ci.yml:77]: why are you doing this?

File: langfuse-core/src/index.ts [hunk 2]
```diff
@@ -564,46 +569,54 @@ abstract class LangfuseCoreStateless {
...
-      }
+  protected async processEnqueueEvent(type: LangfuseObject, body: EventBody): Promise<void> {
+    this.maskEventBodyInPlace(body);
+    await this.processMediaInEvent(type, body);
+    const finalEventBody = this.truncateEventBody(body, MAX_EVENT_SIZE);
+
+    try {
+      JSON.stringify(finalEventBody);
+    } catch (e) {
+      throw Error(`[Langfuse SDK] Event Body for ${type} is not JSON-serializable: ${e}`);
```
Comment (maxdeichmann) [inline langfuse-core/src/index.ts:597]: Why did you decide to throw in this function? The callers should all catch exceptions, but this was another security layer to never throw in the critical path of our users.

File: langfuse-core/src/index.ts [hunk 3]
```diff
@@ -674,14 +687,207 @@ abstract class LangfuseCoreStateless {
     }
   }
 
+  protected async processMediaInEvent(type: LangfuseObject, body: EventBody): Promise<void> {
+    if (!body) return;
+
+    const traceId = "traceId" in body ? body.traceId : type.includes("trace") ? body.id : undefined;
+
+    if (!traceId) {
+      throw new Error("traceId is required for media upload");
```
Comment (maxdeichmann) [inline langfuse-core/src/index.ts:696]: Whya re you throwing here? I would just return as this is not an error condition, right?

File: langfuse-core/src/index.ts [hunk 4]
```diff
@@ -674,14 +687,207 @@ abstract class LangfuseCoreStateless {
...
+        return media;
+      }
+      if (typeof data !== "object" || data === null) {
+        return data;
+      }
+
+      // Use WeakMap to detect cycles
+      if (seenObjects.has(data) || level > maxLevels) {
+        return data;
+      }
```
Comment (maxdeichmann) [inline langfuse-core/src/index.ts:746]: should this not be at the top of the function?

File: langfuse-core/src/index.ts [hunk 5]
```diff
@@ -674,14 +687,207 @@ abstract class LangfuseCoreStateless {
...
+          this._getFetchOptions({ method: "PATCH", body: JSON.stringify(patchMediaBody) })
+        );
+        this._events.emit("debug", `Media upload status reported for ${mediaId}`);
+      } else {
+        this._events.emit("debug", `Media ${mediaId} already uploaded`);
+      }
+    } catch (err) {
+      this._events.emit("error", `Error processing media item: ${err}`);
+    }
+  }
```
Comment (maxdeichmann) [inline langfuse-core/src/index.ts:877]: for all fetch() calls in this function, what happens on errors / wrong status codes?

I assume that uploadUrlResponse in `const uploadUrlResponse = (await fetchResponse.json()) as GetMediaUploadUrlResponse;` will for example contain wrong data.

File: langfuse-core/src/index.ts [hunk 6]
```diff
@@ -564,46 +569,54 @@ abstract class LangfuseCoreStateless {
...
+      .finally(() => {
+        delete this.pendingEventProcessingPromises[promiseId];
+      });
+  }
 
-        return;
-      }
+  protected async processEnqueueEvent(type: LangfuseObject, body: EventBody): Promise<void> {
+    this.maskEventBodyInPlace(body);
+    await this.processMediaInEvent(type, body);
```
Comment (maxdeichmann) [inline langfuse-core/src/index.ts:591]: Here, we create the promise to upload when we add an event to the ingestion queue. This is a bit unexpected. I would have expected this to happen in the logic in the flush function.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (hassiebp) [reply .github/workflows/ci.yml:77]: we need the s3 setup if we want to test multimodal in SDK

File: langfuse-core/src/index.ts (hunk 4 above)
Comment (hassiebp) [reply langfuse-core/src/index.ts:746]: weakmap does not support primitives, so need to handle primitives first (we don't recurse them anyways) prior to handling objects

File: langfuse-core/src/index.ts (hunk 5 above)
Comment (hassiebp) [reply langfuse-core/src/index.ts:877]: The fetch api throws only on network errors (timeout, connection failure, dns lookup...) but not if status code is 4xx / 5xx . It exposes a `ok` method to check whether range is 2xx.

I have not yet implemented retries, that will be a follow up task as I have to budget time

**Pull request merged**